### ISSUE TITLE
Joining group with large number of members kills server

### ISSUE THREAD

**Issue by duncaninnes** [uyuni-project/uyuni#2405]
Indexing seems missing when trying to add/remove systems from very large groups.

I'm speaking from the Red Hat Satellite branch of Spacewalk, so hats off if you've already fixed this.  But we found a problem where joining a system to a group during activation key at registration time took longer depending on how many systems were already in that group.  Customer has > 15k systems in the group in question and adding a system to this group takes a long time.  Adding a system to a group with fewer members is much less of an issue.  It's almost as if the index which identifies the systems in a group is missing.  Or a search which is being used is badly formed and not using an appropriate index.

Our solution was to join all systems to a temporary group during registration, then run a batch job to migrate any systems from the temporary group to the real group.

This has a significant effect on performance when used with large numbers of systems and adding/removing systems from large groups.

Hope this helps your project.

**Comment by duncaninnes** [comment]
Further description:

Activation key AK_TestA joins a system to GroupA during registration (and no other groups).
Activation key AK_TestB joins a system to GroupB during registration (and no other groups).
Activation key AK_NoGroup doesn't join a system to any group during registration.

The Activations keys have no other differences (channels, config, etc) - just a difference in the group join.

GroupA already contains 10,000 systems.
GroupB already contains 1000 systems.

We found that registering using AK_NoGroup would be the fastest every time.
We found that registering using AK_TestB would increase the registration time by around 50%.
We found that registering using AK_TestA would increase the registration time by so much that client connections would sometimes time-out.

Looking into this further, doing group joins and leave actions via the WebUI would also show differences in times depending on the group being joined/left.  The API shows similar issues.  It's had such a big impact on performance that the user has had to split the huge group down into 10 different groups to keep performance at a reasonable level.

Running spacewalk-schema-2.5.1-64 if that helps pin the schema down.

**Comment by moio** [comment]
@duncaninnes, thanks for the precious advice!

We did some work in that area but I am not sure we nailed down this particular aspect completely. In particular, I remember of a recent bug report along the same lines, so it is possible the problem is still around.

I will be back to you if we have news!

**Comment by Shirocco88** [comment]
@duncaninnes : Uyuni-2021.05 was released;
could you please check if this issue is fixed with Uyuni-2021.05 ? can we close this issue ?

**Comment by duncaninnes** [comment]
Sorry - I have no ability to test this with Uyuni.  I posses neither a Uyuni installation, or the number of clients to be able to test this.  I only raised the issue as I noticed the performance hit on Red Hat Satellite 5.  Since Satellite is EOL and Spacewalk looks dead, your code base was the only one I could raise the issue against.

I raised the issue in the hope that Uyuni can fix the problem and it results in a smoother path for your users (especially the large ones).

Good Luck.
